Sources/SwiftKueryMySQL/MySQLConnection.swift

```diff
@@ -229,79 +263,160 @@ public class MySQLConnection: Connection {
-    /// - Throws: QueryError.syntaxError if query build fails, or a database error if it fails to prepare statement.
-    public func prepareStatement(_ raw: String) throws -> PreparedStatement  {
-        return try MySQLPreparedStatement(raw, mysql: mysql)
+    /// - Parameter onCompletion: The function to be called when the statement has been prepared.
+    private func prepareStatement(_ raw: String, query: Query? = nil, onCompletion: @escaping ((QueryResult) -> ())) {
+        DispatchQueue.global().async {
+            mysql_thread_init()
+
+            guard let mysql = self.mysql else {
+                return self.runCompletionHandler(.error(QueryError.connection("Connection not connected")), onCompletion: onCompletion)
+            }
+
+            guard let statement = mysql_stmt_init(mysql) else {
+                let error = self.getError(mysql)
+                mysql_thread_end()
+                return self.runCompletionHandler(.error(QueryError.databaseError(error)), onCompletion: onCompletion)
+            }
+
+            guard mysql_stmt_prepare(statement, raw, UInt(raw.utf8.count)) == 0 else {
+                let error = "ERROR \(mysql_stmt_errno(statement)): " + String(cString: mysql_stmt_error(statement))
+                mysql_stmt_close(statement)
+                mysql_thread_end()
+                return self.runCompletionHandler(.error(QueryError.databaseError(error)), onCompletion: onCompletion)
+            }
+
+            let stmt = MySQLPreparedStatement(query: query, mysql: mysql, statement: statement)
+            mysql_thread_end()
+            return self.runCompletionHandler(.success(stmt), onCompletion: onCompletion)
+        }
     }
 
     /// Release a prepared statement.
     ///
     /// - Parameter preparedStatement: The prepared statement to release.
     /// - Parameter onCompletion: The function to be called when the execution has completed.
     public func release(preparedStatement: PreparedStatement, onCompletion: @escaping ((QueryResult) -> ())) {
-        (preparedStatement as! MySQLPreparedStatement).release()
-        onCompletion(.successNoData)
+        DispatchQueue.global().async {
+            mysql_thread_init()
+            let mysqlStmt = preparedStatement as! MySQLPreparedStatement
+            mysqlStmt.release { result in
+                self.runCompletionHandler(result, onCompletion: onCompletion)
+                return
+            }
+            mysql_thread_end()
+        }
     }
 
     /// Execute a prepared statement.
     ///
     /// - Parameter preparedStatement: The prepared statement to execute.
     /// - Parameter onCompletion: The function to be called when the execution has completed.
     public func execute(preparedStatement: PreparedStatement, onCompletion: @escaping ((QueryResult) -> ()))  {
-        (preparedStatement as! MySQLPreparedStatement).execute(onCompletion: onCompletion)
+        let mysqlStmt = preparedStatement as! MySQLPreparedStatement
```

Comment:
Won't this blow up if someone passes in something that isn't a `MySQLPreparedStatement`? It might be better to use a `guard let` and return an appropriate error within the QueryResult...

Reply:
Added handling of the error case.